concordium-consensus/tests/scheduler/SchedulerTests/SmartContracts/V1/CrossMessaging.hs

```diff
@@ -0,0 +1,125 @@
+
+import Test.Hspec
+import Test.HUnit(assertFailure, assertEqual)
+
+import qualified Data.ByteString.Short as BSS
+import qualified Data.ByteString as BS
+import Data.Serialize(runPut, putWord64le, putByteString, putWord16le)
+import Lens.Micro.Platform
+
+import qualified Concordium.Scheduler.Types as Types
+import Concordium.Scheduler.Runner
+
+import Concordium.GlobalState.Instance
+import Concordium.GlobalState.Basic.BlockState.Accounts as Acc
+import Concordium.GlobalState.Basic.BlockState.Instances
+import Concordium.GlobalState.Basic.BlockState
+import Concordium.Wasm
+
+import Concordium.Scheduler.DummyData
+import Concordium.GlobalState.DummyData
+import Concordium.Types.DummyData
+import Concordium.Crypto.DummyData
+
+import SchedulerTests.TestUtils
+
+
+initialBlockState :: BlockState PV4
+initialBlockState = blockStateWithAlesAccount
+    100000000
+    (Acc.putAccountWithRegIds (mkAccount thomasVK thomasAccount 100000000) Acc.emptyAccounts)
+
+counterSourceFile :: FilePath
+counterSourceFile = "./testdata/contracts/v1/call-counter.wasm"
+
+version1 :: WasmVersion
+version1 = V1
+
+proxySourceFile :: FilePath
+proxySourceFile = "./testdata/contracts/v1/send-message-v1.wasm"
+
+version0 :: WasmVersion
+version0 = V0
+
+-- This test sets up two contracts. The counter contract on address 0,0 and the
+-- proxy contract on address 1,0. Then it invokes a single method on the counter
+-- contract. That method calls the forward method on the proxy contract which
+-- forwards the call to the inc method of the counter contract, which finally
+-- increments the counter.
+testCases :: [TestCase PV4]
+testCases =
+  [ TestCase
+    { tcName = ""
+    , tcParameters = defaultParams {tpInitialBlockState=initialBlockState}
+    , tcTransactions =
+      [ ( TJSON { payload = DeployModule version1 counterSourceFile
+                , metadata = makeDummyHeader alesAccount 1 100000
+                , keys = [(0,[(0, alesKP)])]
+                }
+        , (SuccessWithSummary ensureSuccess, emptySpec)
+        )
+      , ( TJSON { payload = DeployModule version0 proxySourceFile
+                , metadata = makeDummyHeader alesAccount 2 100000
+                , keys = [(0,[(0, alesKP)])]
+                }
+        , (SuccessWithSummary ensureSuccess, emptySpec)
+        )
+      , ( TJSON { payload = InitContract 0 version1 counterSourceFile "init_counter" ""
+                , metadata = makeDummyHeader alesAccount 3 100000
+                , keys = [(0,[(0, alesKP)])]
+                }
+        , (SuccessWithSummary ensureSuccess, counterSpec 0)
+        )
+      , ( TJSON { payload = InitContract 0 version0 proxySourceFile "init_proxy" ""
+                , metadata = makeDummyHeader alesAccount 4 100000
+                , keys = [(0,[(0, alesKP)])]
+                }
+        , (SuccessWithSummary ensureSuccess, emptySpec)
+        )
+      , ( TJSON { payload = Update 0 (Types.ContractAddress 0 0) "counter.inc10nocheck" callArgs
```

Comment:
Yes, the check one checks that the return value matches the state, but V0 contracts have no return value. So this would fail with the `inc10` one. I added a comment.